### ISSUE TITLE
Backward hook is not registered for traced modules 

### ISSUE THREAD

**Issue by tugsbayasgalan** [pytorch/pytorch#56935]
## 🐛 Bug

Internal user tried to register backward hook on module parameter but seems like it works on eager mode while failing on traced mode. 

## To Reproduce

```
import torch

class Module(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.logged_data = {}
        self.w = torch.nn.Parameter(torch.tensor(3.0))
        self.w.register_hook(self.get_backward_hook())

    def get_backward_hook(self):
        def backward_hook(grad):
            self.logged_data["grad"] = grad
        return backward_hook

    def forward(self):
        res = {"loss": self.w*2}
        res.update(self.logged_data)
        self.logged_data = {}
        return res

m = Module()
eager_out = m()
print(eager_out)
eager_out["loss"].backward()
print("EAGER RESULT: ", m())

traced_m = torch.jit.trace(m, [], strict=False)
traced_out = traced_m()
print("TRACED RESULT: ", traced_out)
```

Following snippet returns:

```
EAGER RESULT:  {'loss': tensor(6., grad_fn=<MulBackward0>), 'grad': tensor(2.)}
TRACED RESULT:  {'loss': tensor(6., grad_fn=<MulBackward2>)}
```

## Expected behavior

The results should match 

cc @gmagogsfm

**Comment by Lilyjjo** [comment]
I commented on the internal post; JIT doesn't have support for backward hooks. See this issue for tracking: https://github.com/pytorch/pytorch/issues/34329

**Comment by Tihaw** [comment]
The same problem for my Jit Scripted module...

**Comment by Tihaw** [comment]
ChatGPT suggests the gradients can be get via autograd instead of using backward hooks. Still working on this, hope it helps.
